System services

tomas-langer opened the pull request:
Signed-off-by: Tomas Langer <tomas.langer@oracle.com>

Resolves #780

romain-grecourt commented:
If we make `SystemService` extend `Service` then we could update the webserver services for the media support in a backward compatible way. I thought you were going to do that, any reason you did not ?

spericas commented:
As I stated in the other issue, I'm a bit concerned that we are only thinking about these systems services as _request processors_ (actually "processor" may be a better name too). Cross-cutting concerns in system like Helidon should also, at least, include support for _response processors_. Are we only seeing _half_ of the picture?

batsatt commented on docs/src/main/docs/webserver/11_system-services.adoc:
```diff
@@ -0,0 +1,71 @@
+server:
+  system-services:
+    access-log:
+        fromat: "common"
```
typo: fromat --> format

batsatt commented on webserver/webserver/src/main/java/io/helidon/webserver/RequestRouting.java:
```diff
@@ -100,13 +110,223 @@ public void route(BareRequest bareRequest, BareResponse bareResponse) {
-            Contexts.runInContext(nextRequests.context(), (Runnable) nextRequests::next);
+
+            // we must execute system services
+            Contexts.runInContext(nextRequests.context(), () -> {
```
If it will be common to have zero system services, perhaps this whole block should be inside a conditional to avoid adding request overhead.

tomas-langer replied:
There is one line that is always executed `nextRequests.next()`

batsatt replied:
```
List<SystemService> systemServices = webServer.configuration().systemServices();
if (systemServices.isEmpty()) {
    Contexts.runInContext(nextRequests.context(), (Runnable) nextRequests::next);
} else {
    Contexts.runInContext(nextRequests.context(), () -> {
        ServerRequest systemRequest = systemRequest(nextRequests);
        ServerResponse systemResponse = systemResponse(response);
        for (SystemService service : systemServices) {
            service.handle(systemRequest, systemResponse);
        }
        nextRequests.next();
    });
}
```
I understand that the extra complexity is not ideal (and certainly unwarranted if the list will generally be non-empty), but I do worry about unnecessary overhead in request handling slowing us down.

tomas-langer replied:
ok

tomas-langer replied:
There is no difference in performance between your proposed solution and my PR. You are doing exactly the same decision, only outside of the `Context.runInContext`

tomas-langer commented:
Extending a `Service` was originally planned, but then we (during our meeting) said that this should be executed with each and every requests, and as such it should not really support registering additional routes.
You should be able to stuff in the `handle` method that has access to request and response and can register appropriate readers and writers.
I am not sure about @spericas comment - you have access to both request and response, so what is the use case you mean?

spericas commented:
@tomas-langer Can you access a response _after_ it has been updated by the application's handler?

tomas-langer commented:
@spericas I understand. No you cannot do that with current design

tomas-langer commented:
I have added support for request and response processing.
I think the name `SystemService` is not good, as it implies operating system services. We should come up with a better one. Maybe `WebServerExtension`?
@batsatt I have modified `send` methods in WebServer as they were doing the same thing. This should not impact performance (maybe in a positive way, as the empty publisher no longer goes through reactive streams adapter)

batsatt commented:
Agreed that `SystemService` is not a good name. Jersey uses the term "filter", though it separates out requests and responses by class: `ContainerRequestFilter` and `ContainerResponseFilter`. 

This seems like a good term so long as it is possible for an implementation of `processRequest()` or `processResponse()` to throw an exception to terminate the request. 

Maybe `WebFilter` or `WebServerFilter`?

spericas commented:
I tend to agree with @batsatt. These are basically filters, and probably _should_ be allowed to terminate processing, exceptionally or not. When we had this discussion, I brought authentication as an example of early termination. 

@tomas-langer What's the rationale for not supporting those use cases? 

On the subject of interfaces, having single-method interfaces enables the use of lambdas, but that may not be a significant benefit here.

romain-grecourt commented:
Is such WebFilter any different than a normal Handler ?

tomas-langer commented:
I did not consider security as a use case for these, as the security configuration is path base - I actually need to register it per path.
Nevertheless it would be quite good to know it is going to be executed before the business logic...

The complexity would be quite high though - 
 1. a request filter (can terminate, registered for each and every path and method)
 2. a response filter (called right before data is written to the response)
 3. a web server context enhancer (update application context)
 4. a Service that can add processing per method and path

If we allow termination, we need to support request.next() - I wanted to avoid that as well.

If you want to allow all above, we should probably end up with 
 1. A Service - this allows registration per path and method for each request
 2. A "response" service - this allows registrations as above, only invoked before send (after business logic processing)
 3. An extension that can configure application wide context

We could probably start with 1. and 3., and see if there is a use case for 2.

tomas-langer commented:
Also there is already a "filter" in WebServer response, which is a DataChunk Processor

romain-grecourt commented:
Register the system services against in ServerConfiguration feels odd. How about doing it directly within the WebServer builder ?

tomas-langer commented:
Removing this pull request as this is a complex issue that we should handle for version 2.0.
I will update the issue.
